buildSrc/src/main/java/org/elasticsearch/gradle/ReaperPlugin.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle;
+
+import org.gradle.BuildListener;
+import org.gradle.BuildResult;
+import org.gradle.api.GradleException;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.initialization.Settings;
+import org.gradle.api.invocation.Gradle;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+/**
+ * A plugin to handle reaping external services spawned by a build if Gradle dies.
+ */
+public class ReaperPlugin implements Plugin<Project> {
+
+    private Project project;
+    private Path baseDirectory;
+    private volatile Process reaperProcess;
+
+    @Override
+    public void apply(Project project) {
+        this.project = project;
+        this.baseDirectory = project.getRootDir().toPath().resolve(".gradle")
+            .resolve("reaper").resolve("build-" + ProcessHandle.current().pid());
+        project.getGradle().addListener(new BuildListener() {
+            @Override
+            public void buildStarted(Gradle gradle) {}
+
+            @Override
+            public void settingsEvaluated(Settings settings) {}
+
+            @Override
+            public void projectsLoaded(Gradle gradle) {}
+
+            @Override
+            public void projectsEvaluated(Gradle gradle) {}
+
+            @Override
+            public void buildFinished(BuildResult result) {
+                if (reaperProcess != null) {
+                    try {
+                        reaperProcess.getOutputStream().close();
```

Comment:
As mentioned below, I think this might cause funkiness in some scenarios as we are effectively closing the Gradle process inputstream here. How does this even work for long-lived daemons? Does this get reinitialized on every build? On a non-daemon build what happens? Probably best to control this ourselves rather than rely on implicit behavior.

Reply:
This is the key of how the reaper works. It waits on stdin for a read. That call will not complete until the pipe with the parent process (here) is broken. In this case, we are explicitly closing since we had a normally completing build (gradle did not die). It is my understanding Plugin instances are not reused across builds within the same daemon. So whether or not the daemon is long lived, this should signal the reaper at the end of the build to shutdown. What implicit behavior does it look like we are relying on?

Reply:
That comment was based on my assumption that the `ProcessBuilder` passed along the parent's input directly, rather than via a pipe so I think we are ok here in that we will never call `close()` on the same stream twice.